lib/trees/tree.nit

```diff
@@ -0,0 +1,614 @@
+		end
+		return nodes
+	end
+
+	# Nodes in post-order sequence
+	#
+	# Return a list of nodes contained in `self` following a post-order sequence.
+	#
+	# See https://en.wikipedia.org/wiki/Tree_traversal#Post-order.
+	#
+	# Examples:
+	# ~~~
+	# var tree = new Tree[Int, String].from_string("""
+	# a
+	#  b
+	#   d
+	#  c
+	#   e
+	#    g
+	#   f
+	#    h
+	#    i""")
+	# assert tree.postorder.to_s == """[{d},{b},{g},{e},{h},{i},{f},{c},{a}]"""
+	#
+	# tree = new Tree[Int, String].from_string("""
+	# 6
+	#  5
+	#   2
+	#    1
+	#   3
+	#   4""")
+	# assert tree.postorder.to_s == """[{1},{2},{3},{4},{5},{6}]"""
+	# ~~~
+	fun postorder: Array[N] do
+		var root = self.root
+		var nodes = new Array[N]
+		if root == null then return nodes
+		root.recursive_postorder nodes
+		return nodes
+	end
+
+	# TODO comment and test
+	# Compute the Tree Edition Distance between `self` and `o`.
+	#
+	# Tree distance is computed based on the minimum sequence of edit operation
+	# to go from a tree to another.
+	# Edit operations are insert, remove and update.
+	# Each operation is associated to cost that can be modified with `insert_cost`,
+	# `remove_cost` and `update_cost`.
+	#
+	# Two trees are considered similar if their edit distance if equals to 0.
+	# The greater the distance is, greater are the differences between the trees.
+	#
+	# Examples:
+	# ~~~
+	# var tree1 = new Tree[Int, String].from_string("""
+	# a
+	#  b
+	#  c""")
+	# var tree2 = new Tree[Int, String].from_string("""
+	# a
+	#  b
+	#  c""")
+	# assert tree1.distance_from(tree2) == 0
+	# ~~~
+	# Since `tree1` and `tree2` are identical, their distance is 0.
+	#
+	# ~~~
+	# tree1 = new Tree[Int, String].from_string("""
+	# a
+	#  b
+	#  c""")
+	# tree2 = new Tree[Int, String].from_string("""
+	# a
+	#  b
+	#  e
+	#   c""")
+	# assert tree1.distance_from(tree2) == 1
+	# ~~~
+	# Here it takes 1 insertion of `e` to go from `tree1` to `tree2`.
+	#
+	# ~~~
+	# tree1 = new Tree[Int, String].from_string("""
+	# a
+	#  b
+	#   e
+	#    f
+	#   d
+	#   c""")
+	# tree2 = new Tree[Int, String].from_string("""
+	# a
+	#  e
+	#   f
+	#  c
+	#   d
+	#  b""")
+	# assert tree1.distance_from(tree2) == 3
+	# ~~~
+	# `tree1` can be transformed in `tree2` with 3 operations:
+	# 1. remove `b`
+	# 2. update `c` into `b`
+	# 3. insert `c` as parent of `d`
+	#
+	# ~~~raw
+	#             Step 1:      Step 2:      Step 3:
+	#    a           a            a            a
+	#    |         / | \        / | \        / | \
+	#    b        e  d  c      e  d  b      e  c  b
+	#  / | \      |            |            |  |
+	# e  d  c     f            f            f  d
+	# |
+	# f
+	# ~~~
+	#
+	# See:
+	# > *Kaizhong Zhang and Dennis Shasha.*
+	# > Simple fast algorithms for the editing distance between trees and
+	# > related problems.
+	# > SIAM Journal of Computing, 18:1245–1262, 1989.
+	fun distance_from(o: SELF): Int do
+		# Fill dynamic result map with zeros
+		var tdist = new Array[Array[Int]]
```

Comment:
A simple Matrix class can be added to the stdlib. If you want an efficient one like what Matlab does, this require more work.